lib/src/cache/cache.dart

```diff
@@ -0,0 +1,245 @@
+
+    // Await any pending cached futures
+    if (_cache.containsKey(id)) {
+      return _cache[id].then((TValue value) async {
+        await onGet(id, value);
+        return value;
+      });
+    }
+
+    // Install Future value
+    final Completer<TValue> completer = new Completer<TValue>();
+    _cache[id] = completer.future;
+    final TValue value = valueFactory.call();
+
+    onGet(id, value).then((Null _) {
+      _didUpdateController.add(new CacheContext(id, value));
+      completer.complete(value);
+    });
+
+    return _cache[id];
+  }
+
+  /// Returns a value from the cache for a given [TIdentifier] or adds the
+  /// [TValue] resolved from the [Future] returned by factory when the value
+  /// does not exist in the cache.
+  ///
+  /// All calls to [get] await a call to the [onGet] lifecycle method before
+  /// returning the cached value. If the [Cache] does not contain an instance
+  /// for the given [TIdentifier], the given [valueFactory] is called and an
+  /// [CacheContext] event is emitted on the [didUpdate] stream. A call to [get]
+  /// that returns a cached value does not emit this event as the [Cache] has
+  /// not updated.
+  ///
+  /// Calls to [get] are evaluated syncronously and will return the future value
+  /// established by the first call to [get]. This allows calls to get to
+  /// return the same value regardless of completion of asyncronous calls to
+  /// the [onGet] lifecycle methods or [valueFactory].
+  ///
+  /// ```
+  /// var a = cache.getAsync('id', _superLongAsyncCall);
+  /// var b = cache.getAsync('id', _superLongAsyncCall);
+  /// var c = cache.getAsync('id', _superLongAsyncCall);
+  ///
+  /// // All of the values in this list are the same instance returned by the
+  /// // first call to `_superLongAsyncCall`.
+  /// var values = Future.wait([a, b, c]);
+  /// ```
+  ///
+  /// If the [Cache] [isDisposedOrDisposing] then a [StateError] is thrown.
+  @mustCallSuper
+  Future<TValue> getAsync(TIdentifier id, Func<Future<TValue>> valueFactory) {
+    _throwWhenDisposed('getAsync');
+
+    // Await any pending cached futures
+    if (_cache.containsKey(id)) {
+      return _cache[id].then((TValue value) async {
+        await onGet(id, value);
+        return value;
+      });
+    }
+
+    // Install Future value
+    final Completer<TValue> completer = new Completer<TValue>();
+    _cache[id] = completer.future;
+    valueFactory.call().then((TValue value) async {
+      await onGet(id, value);
+      _didUpdateController.add(new CacheContext(id, value));
+      completer.complete(value);
+    });
+
+    return _cache[id];
+  }
+
+  /// Is the given identifier in the [Cache]?
+  ///
+  /// If the [Cache] [isDisposedOrDisposing] then a [StateError] is thrown.
+  bool isCached(TIdentifier id) {
+    _throwWhenDisposed('determine if identifier is cached');
+    return _cache.containsKey(id);
+  }
+
+  /// Allows consumers to define behavior that is executed when a [TValue] is
```

Comment:
For these callbacks I would just say something like "Logic executed when..." instead of the bit about allowing the consumer to do something. See https://github.com/Workiva/w_module/blob/master/lib/src/lifecycle_module.dart#L452 for an example.